File: Kata-IA-refactoring/Kata-IA-refactoring/Services/AssertionActionService.cs
using Kata_IA_refactoring.Models;

namespace Kata_IA_refactoringTests;

public class AssertionActionService : IAssertionActionService
{
    private readonly ILogger<AssertionActionService> _logger;

    public AssertionActionService(ILogger<AssertionActionService> logger)
    {
        _logger = logger;
    }

    public AssertionActionResponse GetAssertionActionResponse(AssertionConsumerServiceRequest assertionConsumerServiceRequest, ClaimsInfos claimsInfos, string requestedEmail)
    {
        if (assertionConsumerServiceRequest.IsClaimsRequired && !ValidateMandatoryClaims(claimsInfos, requestedEmail))
        {
            return new AssertionActionResponse
            {
                Status = AssertionActionStatus.Error,
                Message = "The mandatory claims are not correctly configured in the identity provider"
            };
        }

        return new AssertionActionResponse
        {
            Status = AssertionActionStatus.Success
        };
    }

    private bool ValidateMandatoryClaims(ClaimsInfos claimsInfos, string requestedEmail)
    {
        bool claimsIsValid = true;

        if (string.IsNullOrEmpty(claimsInfos.Email))
        {
            _logger.LogWarning($"The claim {nameof(claimsInfos.Email)} is not correctly configured in the identity provider for this user {requestedEmail}");
            claimsIsValid = false;
        }

        if (string.IsNullOrEmpty(claimsInfos.UserName))
        {
            _logger.LogWarning($"The claim {nameof(claimsInfos.UserName)} is not correctly configured in the identity provider for this user {requestedEmail}");
            claimsIsValid = false;
        }

        if (string.IsNullOrEmpty(claimsInfos.FirstName))
        {
            _logger.LogWarning($"The claim {nameof(claimsInfos.FirstName)} is not correctly configured in the identity provider for this user {requestedEmail}");
            claimsIsValid = false;
        }

        if (string.IsNullOrEmpty(claimsInfos.LastName))
        {
            _logger.LogWarning($"The claim {nameof(claimsInfos.LastName)} is not correctly configured in the identity provider for this user {requestedEmail}");
            claimsIsValid = false;
        }

        return claimsIsValid;
    }
}

File: ts/src/AssertionActionService.ts
import { createLogger, format, transports } from 'winston';
import { IAssertionActionService } from './IAssertionActionService';
export interface AssertionConsumerServiceRequest {
    isClaimsRequired: boolean;
}

export interface ClaimsInfos {
    email?: string | null;
    userName?: string | null;
    firstName?: string | null;
    lastName?: string | null;
    [key: string]: string | null | undefined;
}

enum AssertionActionStatus {
    Success = "Success",
    Error = "Error"
}

export interface AssertionActionResponse {
    status: AssertionActionStatus;
    message?: string;
}

export class AssertionActionService implements IAssertionActionService {
    private logger = createLogger({
        level: 'warn',
        format: format.combine(format.timestamp(), format.json()),
        transports: [
            new transports.File({ filename: 'error.log', level: 'warn' })
        ]
    });

    public getAssertionActionResponse(assertionConsumerServiceRequest: AssertionConsumerServiceRequest, claimsInfos: ClaimsInfos, requestedEmail: string): AssertionActionResponse {
        if (assertionConsumerServiceRequest.isClaimsRequired && this.isAnyMandatoryClaimMissing(claimsInfos, requestedEmail)) {
            return {
                status: AssertionActionStatus.Error,
                message: "The mandatory claims are not correctly configured in the identity provider"
            };
        }

        return {
            status: AssertionActionStatus.Success
        };
    }

    private isAnyMandatoryClaimMissing(claimsInfos: ClaimsInfos, requestedEmail: string): boolean {
        const mandatoryClaims = ['email', 'userName', 'firstName', 'lastName'];
        let isMissingClaim = false;

        for (const claim of mandatoryClaims) {
            if (!claimsInfos[claim]) {
                this.logger.warn(`The claim ${claim} is not correctly configured in the identity provider for this user ${requestedEmail}`);
                isMissingClaim = true;
            }
        }

        return isMissingClaim;
    }
}
